core/src/main/java/com/linecorp/armeria/client/endpoint/RingHashEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,205 @@
+                List<Integer> arr = new ArrayList<>();
+                for (Endpoint endpoint : this.endpoints) {
+                    int weight = endpoint.weight();
+                    arr.add(weight);
+                }
+                int x = binarySearch(arr, sizeOfRing);
+                for (Endpoint endpoint : this.endpoints) {
+                    final int weight = endpoint.weight();
+                    final String host = endpoint.host();
+                    // If weight is 3 and x is 1, place 3 times in the ring
+                    // if weight is 1 and x is 1, place once in the ring
+                    // If weight is 3 and x is 3, place 1 times in the ring
+                    final int count = weight / x;
+                    for (int i = 0; i < count; i++) {
+                        final String weightedHost = host + weight;
+                        final int hash = getXXHash(weightedHost);
+                        ring.put(hash, endpoint);
+                    }
+                }
+            }
+
+            private int getSize(Iterable<Endpoint> endpoints) {
+                int count = 0;
+                for (Endpoint endpoint : endpoints) {
+                    count++;
+                }
+                return count;
+            }
+
+            private int binarySearch(List<Integer> arr, int sz) {
+                Collections.sort(arr);
+                int lt = arr.get(0);
+                int rt = arr.get(arr.size() - 1);
+                while (rt > lt + 1) {
+                    int mid = (lt + rt) / 2;
+                    if (isPossible(mid, arr, sz)) mid = rt;
+                    else mid = lt + 1;
+                }
+                return lt;
+            }
+
+            private boolean isPossible(int x, List<Integer> arr, int sz) {
+                int total = 0;
+                for (int w : arr) {
+                    total += w/x;
+                }
+                return sz >= total;
+            }
+
+            int caculateNumberOfEndpointInTheRing(Iterable<Endpoint> endpoints, int gcd) {
+                int numberOfEndpointInTheRing = 0;
+                for (final Endpoint endpoint : endpoints) {
+                    int weight = endpoint.weight();
+                    int count = weight / gcd;
+                    numberOfEndpointInTheRing += count;
+                }
+                return numberOfEndpointInTheRing;
+            }
+
+            int findGcdInEndpoints(Iterable<Endpoint> endpoints) {
+                int gcd = -1;
+                for (Endpoint endpoint : endpoints) {
+                    int weight = endpoint.weight();
+
+                    // initialize gcd
+                    if (gcd == -1) {
+                        gcd = weight;
+                        continue;
+                    }
+
+                    gcd = gcd(gcd, weight);
+                }
+                return gcd;
+            }
+
+            int gcd(int a, int b) {
```

Comment:
Should we replace this tail recursion with iteration? Modern IDE supports automatic conversion.